File: app/Helpers/PhoneHelper.php
<?php

if (!function_exists('normalize_phone_number')) {
    /**
     * Normalize phone number to standard format
     * Handles +255, 0, and 255 prefixes for Tanzania
     */
    function normalize_phone_number($phone)
    {
        // Remove all non-digit characters except +
        $phone = preg_replace('/[^0-9+]/', '', $phone);
        
        // If it starts with +255, convert to 255
        if (strpos($phone, '+255') === 0) {
            return '255' . substr($phone, 4);
        }
        
        // If it starts with 0, convert to 255
        if (strpos($phone, '0') === 0) {
            return '255' . substr($phone, 1);
        }
        
        // If it starts with 255, return as is
        if (strpos($phone, '255') === 0) {
            return $phone;
        }
        
        // If it's a 9-digit number (Tanzania mobile), add 255 prefix
        if (strlen($phone) === 9) {
            return '255' . $phone;
        }
        
        // Return as is if it doesn't match any pattern
        return $phone;
    }
}

if (!function_exists('format_phone_for_display')) {
    /**
     * Format phone number for display
     */
    function format_phone_for_display($phone)
    {
        $normalized = normalize_phone_number($phone);
        
        // If it's a Tanzania number (starts with 255 and has 12 digits)
        if (strpos($normalized, '255') === 0 && strlen($normalized) === 12) {
            return '+255 ' . substr($normalized, 3, 2) . ' ' . substr($normalized, 5, 3) . ' ' . substr($normalized, 8, 4);
        }
        
        return $phone;
    }
}

if (!function_exists('validate_tanzania_phone')) {
    /**
     * Validate Tanzania phone number
     */
    function validate_tanzania_phone($phone)
    {
        $normalized = normalize_phone_number($phone);
        
        // Tanzania mobile numbers should be 12 digits starting with 255
        if (strlen($normalized) === 12 && strpos($normalized, '255') === 0) {
            return true;
        }
        
        return false;
    }
}

if (!function_exists('find_user_by_phone')) {
    /**
     * Find user by phone number with flexible matching
     */
    function find_user_by_phone($phone)
    {
        $normalized = normalize_phone_number($phone);
        
        // Try to find user with normalized phone number
        $user = \App\Models\User::where('phone', $normalized)->first();
        
        if ($user) {
            return $user;
        }
        
        // If not found, try different variations
        $variations = [];
        
        // If it's a Tanzania number, try different formats
        if (strpos($normalized, '255') === 0 && strlen($normalized) === 12) {
            $number = substr($normalized, 3); // Remove 255 prefix
            
            $variations = [
                $normalized,                    // 255xxxxxxxxx
                '0' . $number,                  // 0xxxxxxxxx
                '+' . $normalized,              // +255xxxxxxxxx
                $number                         // xxxxxxxxx (9 digits)
            ];
        }
        
        // Try each variation
        foreach ($variations as $variation) {
            $user = \App\Models\User::where('phone', $variation)->first();
            if ($user) {
                return $user;
            }
        }
        
        return null;
    }
} 

if (!function_exists('find_guardian_by_phone')) {
    /**
     * Find guardian by phone number with flexible matching
     */
    function find_guardian_by_phone($phone)
    {
        // School/College module removed - guardian lookup no longer available
        return null;
    }
} 

if (!function_exists('mask_phone_number')) {
    /**
     * Mask phone number for security (shows only last 4 digits)
     */
    function mask_phone_number($phone)
    {
        $normalized = normalize_phone_number($phone);
        
        // If it's a Tanzania number (starts with 255 and has 12 digits)
        if (strpos($normalized, '255') === 0 && strlen($normalized) === 12) {
            return '+255 *** *** ' . substr($normalized, -4);
        }
        
        // For other formats, mask all but last 4 digits
        if (strlen($normalized) > 4) {
            return str_repeat('*', strlen($normalized) - 4) . substr($normalized, -4);
        }
        
        return str_repeat('*', strlen($normalized));
    }
} 

File: mobile_app/lib/utils/phone_helper.dart
/// Phone Number Helper
/// 
/// Normalizes phone numbers to match the backend format
/// Handles Tanzania phone numbers: +255, 0, and 255 prefixes

class PhoneHelper {
  /// Normalize phone number to standard format (255xxxxxxxxx)
  /// 
  /// Handles:
  /// - +255xxxxxxxxx -> 255xxxxxxxxx
  /// - 0xxxxxxxxx -> 255xxxxxxxxx
  /// - 255xxxxxxxxx -> 255xxxxxxxxx (already correct)
  /// - xxxxxxxxx (9 digits) -> 255xxxxxxxxx
  static String normalizePhoneNumber(String phone) {
    // Remove all non-digit characters except +
    String cleaned = phone.replaceAll(RegExp(r'[^0-9+]'), '');
    
    // If it starts with +255, convert to 255
    if (cleaned.startsWith('+255')) {
      return '255${cleaned.substring(4)}';
    }
    
    // If it starts with 0, convert to 255
    if (cleaned.startsWith('0')) {
      return '255${cleaned.substring(1)}';
    }
    
    // If it starts with 255, return as is
    if (cleaned.startsWith('255')) {
      return cleaned;
    }
    
    // If it's a 9-digit number (Tanzania mobile), add 255 prefix
    if (cleaned.length == 9) {
      return '255$cleaned';
    }
    
    // Return as is if it doesn't match any pattern
    return cleaned;
  }
  
  /// Format phone number for display
  static String formatPhoneForDisplay(String phone) {
    String normalized = normalizePhoneNumber(phone);
    
    // If it's a Tanzania number (starts with 255 and has 12 digits)
    if (normalized.startsWith('255') && normalized.length == 12) {
      String number = normalized.substring(3); // Remove 255 prefix
      return '+255 ${number.substring(0, 2)} ${number.substring(2, 5)} ${number.substring(5)}';
    }
    
    return phone;
  }
  
  /// Validate Tanzania phone number
  static bool validateTanzaniaPhone(String phone) {
    String normalized = normalizePhoneNumber(phone);
    
    // Tanzania mobile numbers should be 12 digits starting with 255
    return normalized.length == 12 && normalized.startsWith('255');
  }
  
  /// Clean phone number (remove spaces, dashes, etc.)
  static String cleanPhoneNumber(String phone) {
    return phone.replaceAll(RegExp(r'[^0-9+]'), '');
  }
}

